cpp/src/bitmask/null_mask.cu

```diff
@@ -455,6 +450,117 @@ CUDF_KERNEL void count_set_bits_kernel(bitmask_type const* bitmask,
+ *
+ * @param bitmasks Array of bitmask pointers
+ * @param num_bitmasks Number of bitmasks to process
+ * @param first_bit_index The index (inclusive) of the first bit to count
+ * @param last_bit_index The index (inclusive) of the last bit to count
+ * @param global_count Output array of set bit counts for each bitmask
+ */
+template <size_type block_size>
+CUDF_KERNEL void batch_count_set_bit_kernel(bitmask_type const* const* bitmasks,
+                                            size_type num_bitmasks,
+                                            size_type first_bit_index,
+                                            size_type last_bit_index,
+                                            size_type* global_count)
+{
+  auto const bitmask_idx = static_cast<size_type>(blockIdx.y);
+  if (bitmask_idx >= num_bitmasks) { return; }
+  auto const bitmask = bitmasks[bitmask_idx];
+  if (bitmask == nullptr) { return; }
+
+  constexpr auto const word_size{detail::size_in_bits<bitmask_type>()};
+  auto const first_word_index{word_index(first_bit_index)};
+  auto const last_word_index{word_index(last_bit_index)};
+
+  // Although we use a 2D grid for launching the kernel, elements of each bitmask are indexed by
+  // the first dimension.
+  thread_index_type const tid    = grid_1d::global_thread_id<block_size>();
+  thread_index_type const stride = grid_1d::grid_stride<block_size>();
+  size_type thread_word_index    = tid + first_word_index;
+  size_type thread_count{0};
+
+  // First, just count the bits in all words
+  while (thread_word_index <= last_word_index) {
+    thread_count += cuda::std::popcount(bitmask[thread_word_index]);
+    thread_word_index += stride;
+  }
+
+  // Subtract any slack bits counted from the first and last word
+  // Two threads handle this -- one for first word, one for last
+  if (tid < 2) {
+    bool const first{tid == 0};
+    bool const last{not first};
+
+    size_type bit_index  = (first) ? first_bit_index : last_bit_index;
+    size_type word_index = (first) ? first_word_index : last_word_index;
+
+    size_type num_slack_bits = bit_index % word_size;
+    if (last) { num_slack_bits = word_size - num_slack_bits - 1; }
+
+    if (num_slack_bits > 0) {
+      bitmask_type word = bitmask[word_index];
+      auto slack_mask   = (first) ? set_least_significant_bits(num_slack_bits)
+                                  : set_most_significant_bits(num_slack_bits);
+
+      thread_count -= cuda::std::popcount(word & slack_mask);
+    }
+  }
+
+  using BlockReduce = cub::BlockReduce<size_type, block_size>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  size_type block_count{BlockReduce(temp_storage).Sum(thread_count)};
+
+  if (threadIdx.x == 0) { atomicAdd(&global_count[bitmask_idx], block_count); }
+}
+
+// Count null elements in the specified range of multiple validity bitmasks
+std::vector<size_type> batch_null_count(host_span<bitmask_type const* const> bitmasks,
```

Comment:
The public `null_count` can simply invoke this function, eliminating the need for `count_set_bits_kernel` and relying solely on `batch_count_set_bit_kernel`.